BJS2-69183 Abstraction, abstraction!

dextomi opened the pull request:
Added abstract Character class with basic fields and methods, implemented Warrior and Archer subclasses, added simple test in Main.

dextomi committed:
Added abstract Character class with basic fields and methods, implemented Warrior and Archer subclasses, added simple test in Main.

dextomi committed:
Added abstract Character class with basic fields and methods, implemented Warrior and Archer subclasses, added simple test in Main.

dextomi committed:
renamed setHealth to updateHealth for it to not match setter

20cognosce commented on src/main/java/school/faang/abstraction_abstraction/Archer.java:
```diff
@@ -0,0 +1,12 @@
+
+public class Archer extends Character {
+    public Archer(String name) {
+        super(name, 3, 10, 5);
```
Магические числа, вынести в public static final константы в отдельном классе-конфигурации

20cognosce commented on src/main/java/school/faang/abstraction_abstraction/Warrior.java:
```diff
@@ -0,0 +1,12 @@
+package school.faang.abstraction_abstraction;
+
+public class Warrior extends Character {
```
Аналогично

20cognosce commented on src/main/java/school/faang/abstraction_abstraction/Character.java:
```diff
@@ -0,0 +1,31 @@
+    private String name;
+    protected int strength;
+    protected int agility;
+    protected int intelligence;
```
они все могут быть private final

20cognosce commented on src/main/java/school/faang/abstraction_abstraction/Character.java:
```diff
@@ -0,0 +1,31 @@
+    protected int intelligence;
+    private int health = 100;
+
+    public Character(String name) {
```
конструкторы можно protected

20cognosce commented on src/main/java/school/faang/abstraction_abstraction/Character.java:
```diff
@@ -0,0 +1,31 @@
+    private int health = 100;
+
+    public Character(String name) {
+        this(name, 5, 5, 5);
```
В константы

20cognosce commented on src/main/java/school/faang/abstraction_abstraction/Character.java:
```diff
@@ -0,0 +1,31 @@
+    protected int strength;
+    protected int agility;
+    protected int intelligence;
+    private int health = 100;
```
тоже

20cognosce commented on src/main/java/school/faang/abstraction_abstraction/Main.java:
```diff
@@ -0,0 +1,27 @@
+            if (dextomi.getHealth() == 0) {
+                System.out.println("iseka wins!");
+            }
+            if (iseka.getHealth() == 0) {
```
вот эти все if можно заменить на switch case

а код внутри блоков вынести в отдельные методы под каждую ситуацию

dextomi committed:
Added CharacterConfig and moved character default stats into config and made them constants;
Added currentHealth field to track health updates;
Made code blocks under conditions as separate methods and changed conditions from if statements to switch case, added enum in configs for it

20cognosce commented on src/main/java/school/faang/abstraction_abstraction/Main.java:
```diff
@@ -0,0 +1,45 @@
+        var characterTurn = CharacterConfig.Turn.PLAYER_TURN;
+
+        while (player.getCurrentHealth() > 0 && opponent.getCurrentHealth() > 0) {
+            switch (characterTurn) {
+                case PLAYER_TURN:
+                    playerAttack(player, opponent);
+                    break;
+                case OPPONENT_TURN:
+                    opponentAttack(player, opponent);
+                    break;
+                default:
+                    break;
+            }
```
можно использовать enhanced switch если ты на джаве 17+
case X -> foo();

20cognosce commented on src/main/java/school/faang/abstraction_abstraction/CharacterConfig.java:
```diff
@@ -0,0 +1,21 @@
+import lombok.AllArgsConstructor;
+
+public class CharacterConfig {
+    @AllArgsConstructor
+    public enum Turn {
+        PLAYER_TURN,
+        OPPONENT_TURN
+    }
```
Вот это лишнее
